src/expand_tilde.c

```diff
@@ -0,0 +1,140 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   expand_tilde.c                                     :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: gwolf <gwolf@student.42vienna.com>         +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/07/13 22:15:14 by gwolf             #+#    #+#             */
+/*   Updated: 2023/07/14 19:38:20 by gwolf            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file expand_tilde.c
+ * @brief Functions to handle expansion of tilde.
+ */
+#include	"expander.h"
+
+/**
+ * @brief Find the corresponding replacement for token.
+ *
+ * Depending on the token find env var.
+ * ~: $HOME.
+ * ~+: $PWD.
+ * ~-: $OLDPWD.
+ * If not found increase update position to account for '~' and
+ * return ERR_NOT_FOUND.
+ * If found ft_strdup() the value.
+ * Calc replace.len.
+ *
+ * @param token Contains searched for token.
+ * @param symtab Environment.
+ * @param replace Where to save replace string.
+ * @param pos Current position.
+ * @return t_error SUCCESS, ERR_NOT_FOUND, ERR_MALLOC
+ */
+t_error	ft_get_tilde_replace(t_tok token, t_hashtable *symtab, t_tok *replace, size_t *pos)
+{
+	char		*target;
+	t_env_var	*env_var;
+
+	if (!ft_strcmp(token.str, "~"))
+		target = "HOME";
+	else if (!ft_strcmp(token.str, "~+"))
+		target = "PWD";
+	else if (!ft_strcmp(token.str, "~-"))
+		target = "OLDPWD";
+	env_var = ft_hashtable_lookup(symtab, target, ft_strlen(target));
+	if (!env_var)
+	{
+		(*pos)++;
+		return (ERR_NOT_FOUND);
+	}
+	replace->str = ft_strdup(env_var->value);
+	if (!replace->str)
+		return (ERR_MALLOC);
+	replace->len = ft_strlen(replace->str);
+	return (SUCCESS);
+}
+
+/**
+ * @brief Check for a recognised token.
+ *
+ * At first the string "~" is assigned.
+ * If '+' or '-' is found increase position and reassign
+ * Then a '/' or a '\0' has to be found, else the string is set to NULL.
+ * Also if not found increase the position to jump over the '~'.
+ * Calc the token.len.
+ *
+ * @param input String.
+ * @param pos Current position.
+ * @param token Where to save the token.
+ * @return t_error SUCCESS.
+ */
+t_error	ft_get_tilde_token(char *input, size_t *pos, t_tok *token)
+{
+	size_t	i;
+
+	i = *pos;
+	token->str = "~";
+	i++;
+	if (input[i] == '+')
+	{
+		token->str = "~+";
+		i++;
+	}
+	else if (input[i] == '-')
+	{
+		token->str = "~-";
+		i++;
+	}
```

Comment:
Not quite: i only want to i++ if i find a plus '+' or a minus '-' after the tilde. If there is neither of the two (which is perfectly valid) i don't want i++
I also tried to somehow get rid of the double i++. But since I'm also assigning the correct token I need to know which char exactly i found. I think the alternative would be to check either plus or minus and the check again which one of the two.

Reply:
```
if (input[i] == '+')
    token->str = "~+";
else if (input[i] == '-')
    token->str = "~-";
if (input[i] == '+' || input[i] == '-')
    i++;
```

Like this you could save 4 lines.

Reply:
True! I think I was trying to avoid double checking the char. Changed